README.md
# NAME

IO::Framed - Convenience wrapper for frame-based I/O

# SYNOPSIS

Reading:

    #See below about seed bytes.
    my $iof = IO::Framed->new( $fh, 'seed bytes' );

    #This returns undef if the $in_fh doesn’t have at least
    #the given length (5 in this case) of bytes to read.
    $frame = $iof->read(5);

    #Don’t call this after an incomplete read().
    $line_or_undef = $iof->read_until("\x0a");

Writing, unqueued (i.e., for blocking writes):

    #The second parameter (if given) is executed immediately after the final
    #byte of the payload is written. For blocking I/O this happens
    #before the following method returns.
    $iof->write('hoohoo', sub { print 'sent!' } );

Writing, queued (for non-blocking writes):

    $iof->enable_write_queue();

    #This just adds to a memory queue:
    $iof->write('hoohoo', sub { print 'sent!' } );

    #This will be 1, since we have 1 message/frame queued to send.
    $iof->get_write_queue_count();

    #Returns 1 if it empties out the queue; 0 otherwise.
    #Partial frame writes are accommodated; the callback given as 2nd
    #argument to write() only fires when the queue item is sent completely.
    my $empty = $iof->flush_write_queue();

You can also use `IO::Framed::Read` and `IO::Framed::Write`, which
contain just the read and write features. (`IO::Framed` is actually a
subclass of them both.)

# DESCRIPTION

While writing [Net::WAMP](https://metacpan.org/pod/Net::WAMP) I noticed that I was reimplementing some of the
same patterns I’d used in [Net::WebSocket](https://metacpan.org/pod/Net::WebSocket) to parse frames from a stream:

- Only read() entire frames, with a read queue for any partials.
- Continuance when a partial frame is delivered.
- Write queue with callbacks for non-blocking I/O
- Signal resilience: resume read/write after Perl receives a trapped
signal rather than throwing/giving EINTR. (cf. [IO::SigGuard](https://metacpan.org/pod/IO::SigGuard))

These are now made available in this distribution.

# ABOUT READS

The premise here is that you expect a given number of bytes at a given time
and that a partial read should be continued once it is sensible to do so.

As a result, `read()` will throw an exception if the number of bytes given
for a continuance is not the same number as were originally requested.
`read_until()` will throw a similar exception if called between an incomplete
`read()` and its completion.

Example:

    #This reads only 2 bytes, so read() will return undef.
    $iof->read(10);

    #… wait for readiness if non-blocking …

    #XXX This die()s because we’re in the middle of trying to read
    #10 bytes, not 4.
    $iof->read(4);

    #If this completes the read (i.e., takes in 8 bytes), then it’ll
    #return the full 10 bytes; otherwise, it’ll return undef again.
    $iof->read(10);

EINTR prompts a redo of the read operation. EAGAIN and EWOULDBLOCK (the same
error generally, but not always) prompt an undef return.
Any other failures prompt an instance of [IO::Framed::X::ReadError](https://metacpan.org/pod/IO::Framed::X::ReadError) to be
thrown.

## End-Match Reads

Reader modules now implement a `read_until()` method, which reads arbitrarily
many bytes until
a given sequence of bytes appears then returns those bytes (plus the looked-for
sequence in the return). An obvious application for this feature
is line-by-line reads, e.g., to implement HTTP or other line-based protocols.

## Empty Reads

This class’s `read()` and `read_until()` methods will, by default, throw
an instance of
[IO::Framed::X::EmptyRead](https://metacpan.org/pod/IO::Framed::X::EmptyRead) on an empty read. This is normal and logical
behavior in contexts (like [Net::WebSocket](https://metacpan.org/pod/Net::WebSocket)) where the data stream itself
indicates when no more data will come across. In such cases an empty read
is genuinely an error condition: it either means you’re reading past when
you should, or the other side prematurely went away.

In some other cases, though, that empty read is the normal and expected way
to know that a filehandle/socket has no more data to read.

If you prefer, then, you can call the `allow_empty_read()` method to switch
to a different behavior, e.g.:

    $framed->allow_empty_read();

    my $frame = $framed->read(10);

    if (length $frame) {
        #yay, we got a frame!
    }
    elsif (defined $frame) {
        #no more data will come in, so let’s close up shop
    }
    else {
        #undef means we just haven’t gotten as much data as we want yet;
        #in this case, that means fewer than 10 bytes are available.
    }

    #----------------------------------------------------------------------
    # The same example as above with line-oriented input …

    my $line = $framed->read_until("\x0a");

    if (length $line) {
        #yay, we got a line!
    }
    elsif (defined $line) {
        #no more data will come in, so let’s close up shop
    }
    else {
        #undef means we just haven’t gotten a full line yet.
    }

Instead of throwing the aforementioned exception, `read()` now returns
empty-string on an empty read. That means that you now have to distinguish
between multiple “falsey” states: undef for when the requested number
of bytes hasn’t yet arrived, and empty string for when no more bytes
will ever arrive. But it is also true now that the only exceptions thrown
are bona fide **errors**, which will suit some applications better than the
default behavior.

NB: If you want to be super-light, you can bring in IO::Framed::Read instead
of the full IO::Framed. (IO::Framed is already pretty lightweight, though.)

# ABOUT WRITES

Writes for blocking I/O are straightforward: the system will always send
the entire buffer. The OS’s `write()` won’t return until everything
meant to be written is written. Life is pleasant; life is simple. :)

Non-blocking I/O is trickier. Not only can the OS’s `write()` write
a subset of the data it’s given, but we also can’t know that the output
filehandle is ready right when we want it. This means that we have to queue up
our writes
then write them once we know (e.g., through `select()`) that the filehandle
is ready. Each `write()` call, then, enqueues one new buffer to write.

Since it’s often useful to know when a payload has been sent,
`write()` accepts an optional callback that will be executed immediately
after the last byte of the payload is written to the output filehandle.

Empty out the write queue by calling `flush_write_queue()` and looking for
a truthy response. (A falsey response means there is still data left in the
queue.) `get_write_queue_count()` gives you the number of queue items left
to write. (A partially-written item is treated the same as a fully-unwritten
one.)

Note that, while it’s acceptable to activate and deactive the write queue,
the write queue must be empty in order to deactivate it. (You’ll get a
nasty, untyped exception otherwise!)

`write()` returns undef on EAGAIN and EWOULDBLOCK. It retries on EINTR,
so you should never actually see this error from this module.
Other errors prompt a thrown exception.

NB: `enable_write_queue()` and `disable_write_queue()` return the object,
so you can instantiate thus:

    my $nb_writer = IO::Framed::Write->new($fh)->enable_write_queue();

NB: If you want to be super-light, you can bring in IO::Framed::Write instead
of the full IO::Framed. (IO::Framed is already pretty lightweight, though.)

# CUSTOM READ & WRITE LOGIC

As of version 0.04, you can override READ and WRITE methods with your
preferred logic. For example, in Linux you might prefer `send()` rather than
`syswrite()` to avoid SIGPIPE, thus:

    package My::Framed;

    use parent qw( IO::Framed::Write );

    #Only these two arguments are given.
    sub WRITE {
        return send( $_[0], $_[1], Socket::MSG_NOSIGNAL );
    }

(NB: In \*BSD OSes you can set SO\_SIGNOPIPE on the filehandle instead.)

You can likewise set `READ()` to achieve the same effect for reads.
(`READ()` receives all four arguments that `sysread()` can consume.)

**IMPORTANT:** Unlike most inherited methods, `READ()` and `WRITE()` do
NOT receive the object instance. They must follow the same semantics as
Perl’s `sysread()` and `syswrite()`: i.e., they must return the number
of bytes read/written, or return undef and set `$!` appropriately on error.

# ERROR RESPONSES

An empty read or any I/O error besides the ones mentioned previously
are indicated via an instance of one of the following exceptions.

All exceptions subclass [IO::Framed::X::Base](https://metacpan.org/pod/IO::Framed::X::Base), which itself
subclasses `X::Tiny::Base`.

- [IO::Framed::X::ReadError](https://metacpan.org/pod/IO::Framed::X::ReadError)
- [IO::Framed::X::WriteError](https://metacpan.org/pod/IO::Framed::X::WriteError)

    These both have an `OS_ERROR` property (cf. [X::Tiny::Base](https://metacpan.org/pod/X::Tiny::Base)’s accessor
    method).

- [IO::Framed::X::EmptyRead](https://metacpan.org/pod/IO::Framed::X::EmptyRead)

    No properties. If this is thrown, your peer has probably closed the connection.
    Unless you have called `allow_empty_read()` to set an alternate behavior,
    you might want to trap this exception if you call `read()`.

**NOTE:** This distribution doesn’t write to `$!`. EAGAIN and EWOULDBLOCK on
`flush_write_queue()` are ignored; all other errors are converted
to thrown exceptions.

# LEGACY CLASSES

This distribution also includes the following **DEPRECATED** legacy classes:

- IO::Framed::Write::Blocking
- IO::Framed::Write::NonBlocking
- IO::Framed::ReadWrite
- IO::Framed::ReadWrite::Blocking
- IO::Framed::ReadWrite::NonBlocking

I’ll keep these in for the time being but eventually **WILL** remove them.
Please adjust any calling code that you might have.

# REPOSITORY

[https://github.com/FGasper/p5-IO-Framed](https://github.com/FGasper/p5-IO-Framed)

# AUTHOR

Felipe Gasper (FELIPE)

# COPYRIGHT

Copyright 2017 by [Gasper Software Consulting, LLC](http://gaspersoftware.com)

# LICENSE

This distribution is released under the same license as Perl.


Changes
Revision history for Perl module IO::Framed

0.16 Sat May 18 2019
    - No production changes.
    - Replace Test::NoWarnings with Test::FailWarnings.

0.15 Thu Dec 6 2018
    - Accommodate IO::SigGuard 0.13’s lazy-loading.

0.14 Fri Aug 24 2018
    - Fix LICENSE file (again)

0.13 Fri Aug 24 2018
    - Fix LICENSE file.

0.12 Fri Aug 24 2018
    - Add read_until() method.

0.11 Sun Aug 12 2018
    - BREAKING CHANGE: Treat zero-length and undefined input to write() as
      failure states.

0.10 Tue Jan 9 2018
    - Switch to two-digit version numbers (rpm seems to like them better)
    - remove README.pod from Makefile target (cf. EU::MM #119)

0.04 Fri Dec 29 2017
    - Allow read methods besides IO::SigGuard::sysread.
    - Allow write methods besides IO::SigGuard::syswrite.
    - Make all errors subclass IO::Framed::X::Base.
    - Supply some missing tests.
    - Resort the changelog in descending chronological order.

0.031 Thu Nov 9 2017
    - Fix up some documentation. No code changes.

0.03 Tue Jul 24 2017
    - Deprecate the ReadWrite class; reassign that functionality to
      the IO::Framed namespace.
    - allow_empty_read() to make read() not throw EmptyRead on an empty read.

0.021 Sun Apr 9 2017
    - Again tweak the tests for MSWin: drop new_from_fd() since it hangs.
      (thank you to CHORNY for pointing out the problem!)

0.02 Wed Apr 5 2017
    - Deprecate the Blocking/NonBlocking write classes. This radically
      simplifies the usage and allows clean switching back and forth between
      blocking (“unqueued”) and non-blocking (“queued”) writes.

0.018 Wed Apr 5 2017
    - Tests: avoid blocking() on normally-created file handles since only
      Perl 5.14+ automatically upgrades file handles to IO::File.

0.017 Wed Apr 5 2017
    - Require the most recent IO::SigGuard in the MakeFile, which
      hopefully will fix some CPAN Testers test reports.

0.016 Wed Apr 5 2017
    - Thanks to CHORNY: Fixes for Strawberry Perl and other OSes that
      treat EWOULDBLOCK as a separate error from EAGAIN.

0.015 Sat Apr 1 2017
    - Fixes for tests on Linux

0.014 Sat Apr 1 2017
    - Add no-op stub queue methods to Write::Blocking so that
      applications don’t need to care whether a given IO object
      is blocking or not.

0.013 Sat Apr 1 2017
    - Fix ReadWrite modules to actually work. :-/

0.012 Sat Apr 1 2017
    - instead of “input”/“output” in accessors,
      use “read”/“write” since those are the module and method names
      and should thus be easier to remember.

0.011 Sat Apr 1 2017
    - Compile/sanity fixes

0.01 Fri Mar 31 2017
    - first release


Makefile.PL

use ExtUtils::MakeMaker;

#hack to accommodate https://github.com/Perl-Toolchain-Gang/ExtUtils-MakeMaker/issues/119
my $old_libscan = \&ExtUtils::MM_Any::libscan;
*ExtUtils::MM_Any::libscan = sub {
    my ($self, $path) = @_;
    return q<> if $path eq 'README.pod';
    return $old_libscan->(@_);
};

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile1(
    NAME         => 'IO::Framed',
    VERSION_FROM => 'lib/IO/Framed.pm',         # finds \$VERSION
    AUTHOR       => 'Felipe Gasper',
    ABSTRACT_FROM => 'lib/IO/Framed.pm',
    LICENSE      => 'perl',

    PREREQ_PM    => {
        'autodie'                  => 0,
        'parent'                   => 0,
        'X::Tiny'   => 0,
        'IO::SigGuard' => 0.013,
    },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources   => {
            bugtracker => 'https://github.com/chansen/p5-IO-Framed/issues',
            repository => {
                type => 'git',
                url  => 'https://github.com/FGasper/p5-IO-Framed.git',
                web  => 'https://github.com/FGasper/p5-IO-Framed',
            },
        },
    },
    TEST_REQUIRES => {
        'Test::More'      => 0,
        'Test::Simple'    => 0,
        'Test::FailWarnings' => 0,
        'Test::Exception' => 0,
        ( $^O eq 'MSWin32' ? ('Win32::Socketpair' => 0.02) : () ),
    },
);

sub WriteMakefile1 {  #Compatibility code for old versions of EU::MM. Written by Alexandr Ciornii, version 2. Added by eumm-upgrade.
    my %params=@_;
    my $eumm_version=$ExtUtils::MakeMaker::VERSION;
    $eumm_version=eval $eumm_version;
    die "EXTRA_META is deprecated" if exists $params{EXTRA_META};
    die "License not specified" if not exists $params{LICENSE};
    if ($params{AUTHOR} and ref($params{AUTHOR}) eq 'ARRAY' and $eumm_version < 6.5705) {
        $params{META_ADD}->{author}=$params{AUTHOR};
        $params{AUTHOR}=join(', ',@{$params{AUTHOR}});
    }
    if ($params{TEST_REQUIRES} and $eumm_version < 6.64) {
        $params{BUILD_REQUIRES}={ %{$params{BUILD_REQUIRES} || {}} , %{$params{TEST_REQUIRES}} };
        delete $params{TEST_REQUIRES};
    }
    if ($params{BUILD_REQUIRES} and $eumm_version < 6.5503) {
        #EUMM 6.5502 has problems with BUILD_REQUIRES
        $params{PREREQ_PM}={ %{$params{PREREQ_PM} || {}} , %{$params{BUILD_REQUIRES}} };
        delete $params{BUILD_REQUIRES};
    }
    delete $params{CONFIGURE_REQUIRES} if $eumm_version < 6.52;
    delete $params{MIN_PERL_VERSION} if $eumm_version < 6.48;
    delete $params{META_MERGE} if $eumm_version < 6.46;
    delete $params{META_ADD} if $eumm_version < 6.46;
    delete $params{LICENSE} if $eumm_version < 6.31;

    WriteMakefile(%params);
}
